File: LinkedList.cpp
#include<stdio.h>
#include<malloc.h>

struct Node
{
	int data;
	struct Node *link;
};

struct Node *head = NULL;

struct Node* createNode()
{
	struct Node *n;
	n = (struct Node*)malloc(sizeof(struct Node));
	return n;
}
void insertAtLast(int item)
{
	struct Node *node,*curr;
	node = createNode();
	node->data = item;
	node->link = NULL;
	if(head==NULL)
	{
		head = node;
	}
	else
	{
		curr = head;
		while(curr->link != NULL)
		{
			curr = curr->link;
		}
		curr->link = node;	
	}	
}
void insertAtFirst(int item)
{
	struct Node *node;
	node = createNode();
	node->data = item;
	node->link = head;
	head = node;
}
void insertAtIndex(int index,int item)
{
	struct Node *node,*curr,*prev;
	node = createNode();
	node->data = item;
	if(index==0)
	{
		node->link = head;
		head = node;
	}
	else
	{
		curr = head;
		for(int i=0;i<index;i++)
		{
			prev = curr;
			curr = curr->link;
		}
		prev->link = node;
		node->link = curr;
	}
}
void print()
{
	struct Node *curr;
	curr = head;
	while(curr->link != NULL)
	{
		printf("%d, ",curr->data);
		curr = curr->link;
	}
	printf("%d",curr->data);
}
bool removeAtIndex(int index)
{
	struct Node *temp,*curr,*prev;
	curr = head;
	int count = 0;
	if(index==0)
	{
		temp = curr->link;
		temp = NULL;
		curr = curr->link;
		head = curr;
		return true;
	}
	else
	{		
		for(int i=0;i<index;i++)
		{
			prev = curr;
			curr = curr->link;
		}
		temp = curr->link;
		prev->link = temp;
		curr = NULL;
		return true;
	}
}
bool removeLast()
{
	struct Node *curr,*prev;
	curr = head;
	while(curr->link != NULL)
	{
		prev= curr;
		curr = curr->link;
	}
	prev->link = NULL;
	curr = NULL;
	return true;
}
void removeFirst()
{
	removeAtIndex(0);
}
void reverse()
{
	struct Node *curr,*prev=NULL,*next;
	curr = head;
	while(curr->link != NULL)
	{
		next = curr->link;
		curr->link =  prev;
		prev = curr;
		curr = next; 
	}
	curr->link = prev;
	head = curr;
}
void deleteValue(int item)
{
	struct Node *curr;
	curr = head;
	int count=0;
	while(curr->link != NULL)
	{
		if(curr->data == item)
		{
			removeAtIndex(count);
			count--;
		}
		count++;
		curr = curr->link;
	}
	if(curr->data == item)
	{
		removeAtIndex(count);
	}
}

void findMiddle()
{
	struct Node *slowPtr=head, *fastPtr=head;
	if(head!=NULL)
	{
		while(fastPtr != NULL && fastPtr->link != NULL)
		{
			fastPtr = fastPtr->link->link;
			slowPtr = slowPtr->link;
		}
		printf("Middle element is : %d\n",slowPtr->data);
	}
	else
	{
		printf("List is Empty");
	}
}

int main()
{
	int choice,item,index;
	bool flag = true;
	while(flag)
	{
		printf("\n1. Insert At Last");
		printf("\n2. Insert At Start");
		printf("\n3. Insert At a particular Index");
		printf("\n4. Remove Last");
		printf("\n5. Remove First");
		printf("\n6. Remove At a particular Index");
		printf("\n7. Reverse");
		printf("\n8. Print");
		printf("\n9. Delete With Value");
		printf("\n10. Find middle element");
		printf("\n11. Exit");
		printf("\nEnter Your Choice");
		scanf("%d",&choice);
		switch(choice)
		{
			case 1:
				scanf("%d",&item);
				insertAtLast(item);
				break;
			case 2:
				scanf("%d",&item);
				insertAtFirst(item);
				break;
			case 3:
				scanf("%d",&index);
				scanf("%d",&item);
				insertAtIndex(index,item);
				break;
			case 4:
				removeLast();
				break;
			case 5:
				removeFirst();
				break;
			case 6:
				scanf("%d",&index);
				removeAtIndex(index);
				break;
			case 7:
				reverse();
				break;
			case 8:
				print();
				break;
			case 9:
				scanf("%d",&item);
				deleteValue(item);
				break;
			case 10:
				findMiddle();
				break;	
			case 11:
				flag = false;
				break;
		}
	}	
	return 0;
}

File: Java/LinkedList/src/LinkedList.java
public class LinkedList
{
	Node head; 
	public void insert(int data)
	{
		Node node = new Node();
		node.data = data;
		//node.next = null;     //it is not necessary. As by default it will be null
		if(head==null)
		{
			head = node;
		}
		else
		{
			Node n = head;
			while(n.next!=null)
			{
				n = n.next;
			}
			n.next = node;
		}	
	}	
	public void show()
	{
		Node n =head;
		while(n.next!=null)
		{
			System.out.println(n.data);
			n = n.next;
		}
		System.out.println(n.data);
	}
	public void insertAtStart(int data)
	{
		Node node = new Node();
		node.data = data;
		node.next = head;
		head = node;
	}
	public void insertAt(int index, int data)
	{
		if(index==0)
		{
			insertAtStart(data);
		}
		else
		{
			Node node = new Node();
			node.data = data;
			Node n = head;
			for(int i=0;i<index-1;i++)
			{
				n = n.next;
			}
			node.next = n.next;
			n.next = node;
		}
	}
	public boolean deleteAt(int index)
	{
		if(index==0)
		{
			head = head.next;
			return true;
		}
		else
		{
			Node n = head;
			for(int i=0;i<index-1;i++)
			{
				n = n.next;
			}
			Node temp = n.next;
			n.next = temp.next;
			temp = null;
			return true;
		}
	}
	public void deleteValue(int value)
	{
		Node n = head;
		int index = 0;
		while(n.next!=null)
		{
			if(n.data == value)
			{
				deleteAt(index);
				index++;
				n = n.next;
			}
			else
			{
				index++;
				n= n.next;
			}
		}
	}
	public void shiftLastToFirst()
	{
		Node n = head;
		Node temp = null;
		while(n.next!=null)
		{
			temp = n;
			n = n.next;
		}		
		temp.next = null;
		n.next = head;
		head = n;
		
	}
	public void delete_k_th_node_from_end(int node_val)
	{
		Node n = head;
		int count = 1;
		while(n.next!=null)
		{
			count++;
			n = n.next;
		}
		int target = count - node_val;
		deleteAt(target);
	}
	public void reverse()
	{
		Node curr = head;
		Node next,prev=null;
		while(curr.next!=null)
		{
			next = curr.next;
			curr.next = prev;
			prev = curr;
			curr =next;
		}
		curr.next = prev;
		head = curr;
	}
	public void removeDuplicate()
	{
		
		
	}
}